Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift

```diff
@@ -0,0 +1,153 @@
+  }
+
+  override func tearDown() {
+    XCTAssertNoThrow(try self.client.close().wait())
+    XCTAssertNoThrow(try self.server.close().wait())
+    XCTAssertNoThrow(try self.group.syncShutdownGracefully())
+    super.tearDown()
+  }
+
+  func testUnary() {
+    XCTAsyncTest {
+      let get = self.echo.makeGetCall(.with { $0.text = "hello" })
+
+      let initialMetadata = try await get.initialMetadata
+      initialMetadata.assertFirst("200", forName: ":status")
+
+      let response = try await get.response
+      XCTAssertEqual(response.text, "Swift echo get: hello")
+
+      let trailingMetadata = try await get.trailingMetadata
+      trailingMetadata.assertFirst("0", forName: "grpc-status")
+
+      let status = await get.status
+      XCTAssertTrue(status.isOk)
+    }
+  }
+
+  func testClientStreaming() {
+    XCTAsyncTest {
+      let collect = self.echo.makeCollectCall()
+
+      try await collect.sendMessage(.with { $0.text = "boyle" })
+      try await collect.sendMessage(.with { $0.text = "jeffers" })
+      try await collect.sendMessage(.with { $0.text = "holt" })
+      try await collect.sendEnd()
+
+      let initialMetadata = try await collect.initialMetadata
+      initialMetadata.assertFirst("200", forName: ":status")
+
+      let response = try await collect.response
+      XCTAssertEqual(response.text, "Swift echo collect: boyle jeffers holt")
+
+      let trailingMetadata = try await collect.trailingMetadata
+      trailingMetadata.assertFirst("0", forName: "grpc-status")
+
+      let status = await collect.status
+      XCTAssertTrue(status.isOk)
+    }
+  }
+
+  func testServerStreaming() {
+    XCTAsyncTest {
+      let expand = self.echo.makeExpandCall(.with { $0.text = "boyle jeffers holt" })
+
+      let initialMetadata = try await expand.initialMetadata
+      initialMetadata.assertFirst("200", forName: ":status")
+
+      let respones = try await expand.responses.map { $0.text }.collect()
+      XCTAssertEqual(respones, [
+        "Swift echo expand (0): boyle",
+        "Swift echo expand (1): jeffers",
+        "Swift echo expand (2): holt",
+      ])
+
+      let trailingMetadata = try await expand.trailingMetadata
+      trailingMetadata.assertFirst("0", forName: "grpc-status")
+
+      let status = await expand.status
+      XCTAssertTrue(status.isOk)
+    }
+  }
+
+  func testBidirectionalStreaming() {
+    XCTAsyncTest {
+      let update = self.echo.makeUpdateCall()
+
+      var responseIterator = update.responses.map { $0.text }.makeAsyncIterator()
+
+      for (i, name) in ["boyle", "jeffers", "holt"].enumerated() {
+        try await update.sendMessage(.with { $0.text = name })
+        let response = try await responseIterator.next()
+        XCTAssertEqual(response, "Swift echo update (\(i)): \(name)")
+      }
+
+      try await update.sendEnd()
+
+      let initialMetadata = try await update.initialMetadata
+      initialMetadata.assertFirst("200", forName: ":status")
```

Comment:
Could we check this before the loop already?

Reply:
At the moment, yes, we could. However, there's no guarantee that initial metadata is sent back on receipt of request metadata. In particular some gRPC implementations offer RPC implementors a way to set initial metadata which is delayed until the server sends a response message. We don't currently do this but it's been requested and I think we will add it in the future. When that's the case then we can't wait for this before sending the first request.

Reply:
Maybe we want to add this as a comment?

Reply:
See #1262 which adds support for our adopters to do just this.

Reply:
Added a comment in 5be83ddbd4ba5c772b3e6317b84c429b87ce859e